docs/upgrading/production-basics.md

```diff
@@ -0,0 +1,233 @@
+---
+sidebar_position: 2
+sidebar_label: Production App Basics
+title: "Upgrading Contracts: Production App Basics"
+---
+
+# Production App Basics
+
+When deploying new code to production contracts, you obviously can't destroy old account state, as you do [during rapid prototyping](./prototyping.md). So how to you prevent the dreaded error?
+
+    Cannot deserialize the contract state.
+
+You can use a couple different approaches, depending on the complexity of your contract.
+
+## Migration method
+
+For cases like [the change to the `rust-status-message` contract](https://github.com/near-examples/rust-status-message/commit/a39e1fc55ee018b631e3304ba6f0884b7558873e) that we looked at [previously](./prototyping.md), a simple migration method is all you need.
+
+As a reminder, the goal was to change this:
+
+```rust reference
+https://github.com/near-examples/rust-status-message/blob/61de649d8311bef5957c129e6ad1407101a0f873/src/lib.rs#L9-L19
+```
+
+into this:
+
+```rust reference
+https://github.com/near-examples/rust-status-message/blob/a39e1fc55ee018b631e3304ba6f0884b7558873e/src/lib.rs#L9-L21
+```
+
+This change seems _almost_ backwards-compatible:
+
+* it renames a field while keeping the [persistent collection](../contract-structure/collections.md) prefix, `r`, the same
+* it adds a new field
+
+In fact, if the contract state was serialized to a self-describing data format such as JSON, it _might_ be backwards-compatible _[TODO: fact check]_, and you could upgrade your contract without encountering any errors. But [Borsh](../contract-interface/serialization-interface.md) needs a little help figuring out how to match the old state to the new contract structure. How to help it?
```

Comment:
plz advise

Reply:
I don't think this is correct unless you are deserializing as an ambiguous type, which isn't ideal. The only benefit of self-describing data formats is if you don't know what the type is you are deserializing, which isn't the case here. You could add an attribute on `serde` to allow a field to not exist and [default it](https://serde.rs/attr-default.html) if you really wanted to lazily migrate state, but the existing keys would need to be the same, which is not the case for the example above. 

This applies the same to borsh as JSON in this case, as you _could_ try to deserialize as each format before handling logic, but this isn't something you would want in the contract long term.

Reply:
Thanks! Now that I'm looking at this again and I know the full details, I don't think this is an aside worth keeping. Do you agree that it makes sense to remove this paragraph and just say something like:

> How do you help your contract load the old Borsh-serialized data into your new contract structure?

Reply:
yeah, something like that sounds good, I personally think it might be a good idea to write documentation around the preference for borsh storage because it's what Rust uses and what AS SDK will be moving towards. I really liked the section where you talked about versioned storage

Reply:
I agree that documentation around "preference for borsh storage" but I don't think the "Upgrading Contracts" docs are the place to get into the details

Reply:
I'm glad you liked the versioned storage bit!